### ISSUE TITLE
segfault in int_ctx_new()

### ISSUE THREAD

**Issue by mdurnev** [OpenSC/libp11#328]
This issue is not reproducible with OpenSSL 1.0.2. I can see it only with OpenSSL 1.1.x.
```
    if (pmeth->init) {
        if (pmeth->init(ret) <= 0) { <== ../openssl-1.1.1d/crypto/evp/pmeth_lib.c:160: It segfaults here, when calling pmeth->init(ret)
            ret->pmeth = NULL;
            EVP_PKEY_CTX_free(ret);
            return NULL;
        }
    }
```
The root cause is that when an engine object is destroyed by ENGINE_free(), pmeth is also cleaned up by calling engine_pkey_meths_free().

The simple workaround is:
```
diff -Naurp a/src/p11_pkey.c b/src/p11_pkey.c
--- a/src/p11_pkey.c    2020-01-28 09:25:11.000000000 +1000
+++ b/src/p11_pkey.c    2020-02-14 21:05:59.568230320 +1000
@@ -666,8 +666,8 @@ int PKCS11_pkey_meths(ENGINE *e, EVP_PKE
                EVP_PKEY_EC,
                0
        };
-       static EVP_PKEY_METHOD *pkey_method_rsa = NULL;
-       static EVP_PKEY_METHOD *pkey_method_ec = NULL;
+       EVP_PKEY_METHOD *pkey_method_rsa = NULL;
+       EVP_PKEY_METHOD *pkey_method_ec = NULL;

        (void)e; /* squash the unused parameter warning */
        /* all PKCS#11 engines currently share the same pkey_meths */
@@ -680,16 +680,14 @@ int PKCS11_pkey_meths(ENGINE *e, EVP_PKE
        /* get the EVP_PKEY_METHOD */
        switch (nid) {
        case EVP_PKEY_RSA:
-               if (pkey_method_rsa == NULL)
-                       pkey_method_rsa = pkcs11_pkey_method_rsa();
+               pkey_method_rsa = pkcs11_pkey_method_rsa();
                if (pkey_method_rsa == NULL)
                        return 0;
                *pmeth = pkey_method_rsa;
                return 1; /* success */
 #ifndef OPENSSL_NO_EC
        case EVP_PKEY_EC:
-               if (pkey_method_ec == NULL)
-                       pkey_method_ec = pkcs11_pkey_method_ec();
+               pkey_method_ec = pkcs11_pkey_method_ec();
                if (pkey_method_ec == NULL)
                        return 0;
                *pmeth = pkey_method_ec;
```
It fixes the issue. But it leaks some memory.
I need a clean solution for this issue.

**Comment by mdurnev** [comment]
This is a backtrace of the moment when the pmeth structure is freed by OpenSSL:
```
(gdb) bt
#0  EVP_PKEY_meth_free (pmeth=0xaaaaaab2b910) at ../openssl-1.1.1d/crypto/evp/pmeth_lib.c:239
#1  0x0000ffffbf301040 in engine_pkey_meths_free (e=e@entry=0xaaaaaab29790) at ../openssl-1.1.1d/crypto/engine/tb_pkmeth.c:110
#2  0x0000ffffbf2fde8c in engine_free_util (e=0xaaaaaab29790, not_locked=<optimized out>) at ../openssl-1.1.1d/crypto/engine/eng_lib.c:87
#3  0x0000ffffbf2fc0a8 in int_engine_configure (cnf=0x0, value=0xaaaaaab064e0 "", name=0xaaaaaab2edb0 "pkcs11") at ../openssl-1.1.1d/crypto/engine/eng_cnf.c:147
#4  int_engine_module_init (md=<optimized out>, cnf=0x0) at ../openssl-1.1.1d/crypto/engine/eng_cnf.c:171
#5  0x0000ffffbf29f41c in module_init (cnf=0xffffbf31e7fc <ossl_init_config_ossl_+28>, value=0xaaaaaab2eb30 "engine_section", name=0xaaaaaab2eae0 "engines", pmod=0xaaaaaaae0840)
    at ../openssl-1.1.1d/crypto/conf/conf_mod.c:312
#6  module_run (flags=50, value=0xaaaaaab2eb30 "engine_section", name=0xaaaaaab2eae0 "engines", cnf=0xffffbf31e7fc <ossl_init_config_ossl_+28>) at ../openssl-1.1.1d/crypto/conf/conf_mod.c:171
#7  CONF_modules_load (cnf=0xffffbf31e7fc <ossl_init_config_ossl_+28>, cnf@entry=0xaaaaaab294b0, appname=appname@entry=0x0, flags=flags@entry=50) at ../openssl-1.1.1d/crypto/conf/conf_mod.c:102
#8  0x0000ffffbf29fa10 in CONF_modules_load_file (filename=filename@entry=0x0, appname=appname@entry=0x0, flags=flags@entry=50) at ../openssl-1.1.1d/crypto/conf/conf_mod.c:138
#9  0x0000ffffbf29fca8 in openssl_config_int (settings=<optimized out>) at ../openssl-1.1.1d/crypto/conf/conf_sap.c:69
#10 0x0000ffffbf31e7fc in ossl_init_config () at ../openssl-1.1.1d/crypto/init.c:293
#11 ossl_init_config_ossl_ () at ../openssl-1.1.1d/crypto/init.c:291
#12 0x0000ffffbf6499f0 in __pthread_once_slow () from target:/lib64/libpthread.so.0
#13 0x0000ffffbf36ba9c in CRYPTO_THREAD_run_once (once=once@entry=0xffffbf424578 <config>, init=init@entry=0xffffbf31e7e0 <ossl_init_config_ossl_>) at ../openssl-1.1.1d/crypto/threads_pthread.c:118
#14 0x0000ffffbf31ee4c in OPENSSL_init_crypto (opts=opts@entry=2097228, settings=0x0) at ../openssl-1.1.1d/crypto/init.c:701
#15 0x0000ffffbf45a26c in OPENSSL_init_ssl (opts=2097228, opts@entry=2097152, settings=settings@entry=0x0) at ../openssl-1.1.1d/ssl/ssl_init.c:205
#16 0x0000ffffbf45dde4 in SSL_CTX_new (meth=0xffffbf4ad3e0 <TLS_client_method_data.23885>) at ../openssl-1.1.1d/ssl/ssl_lib.c:2916
#17 0x0000ffffbf6b06d8 in ossl_connect_step1 () from target:/usr/lib64/libcurl.so.4
#18 0x0000ffffbf6b3938 in ossl_connect_common () from target:/usr/lib64/libcurl.so.4
#19 0x0000ffffbf6b4a10 in Curl_ssl_connect_nonblocking () from target:/usr/lib64/libcurl.so.4
#20 0x0000ffffbf677bdc in https_connecting () from target:/usr/lib64/libcurl.so.4
#21 0x0000ffffbf6794cc in Curl_http_connect () from target:/usr/lib64/libcurl.so.4
#22 0x0000ffffbf693d10 in multi_runsingle () from target:/usr/lib64/libcurl.so.4
#23 0x0000ffffbf694aec in curl_multi_perform () from target:/usr/lib64/libcurl.so.4
#24 0x0000ffffbf68cf54 in curl_easy_perform () from target:/usr/lib64/libcurl.so.4
#25 0x0000aaaaaaabfcc0 in ?? ()
#26 0x0000aaaaaaade000 in ?? ()
```

**Comment by mdurnev** [comment]
And this is a backtrace of the segfault:
```
#0 0x000e0d0c0b0a0908 in ?? ()
#1 0x0000ffffbf31adf8 in int_ctx_new (pkey=0xaaaaaab40f90, pkey@entry=0xaaaaaab3fe60, e=0xaaaaaaaf8d50, e@entry=0x0, id=<optimized out>, id@entry=-1) at ../openssl-1.1.1d/crypto/evp/pmeth_lib.c:160
#2 0x0000ffffbf31aef8 in EVP_PKEY_CTX_new (pkey=pkey@entry=0xaaaaaab3fe60, e=e@entry=0x0) at ../openssl-1.1.1d/crypto/evp/pmeth_lib.c:245
#3 0x0000ffffbf315d90 in do_sigver_init (ctx=ctx@entry=0xaaaaaab4f540, pctx=pctx@entry=0x0, type=type@entry=0xffffbf402b30 <sha256_md>, e=e@entry=0x0, pkey=0xaaaaaab3fe60, pkey@entry=0xaaaaaab40f90,
    ver=ver@entry=1) at ../openssl-1.1.1d/crypto/evp/m_sigver.c:29
#4 0x0000ffffbf315e88 in EVP_DigestVerifyInit (ctx=ctx@entry=0xaaaaaab4f540, pctx=pctx@entry=0x0, type=type@entry=0xffffbf402b30 <sha256_md>, e=e@entry=0x0, pkey=pkey@entry=0xaaaaaab40f90)
    at ../openssl-1.1.1d/crypto/evp/m_sigver.c:97
#5 0x0000ffffbf2582e8 in ASN1_item_verify (it=0xffffbf414488 <X509_CINF_it>, a=a@entry=0xaaaaaab4ecc8, signature=signature@entry=0xaaaaaab4ecd8, asn=asn@entry=0xaaaaaab4ec40, pkey=0xaaaaaab40f90)
    at ../openssl-1.1.1d/crypto/asn1/a_verify.c:148
#6 0x0000ffffbf3846a8 in X509_verify (a=a@entry=0xaaaaaab4ec40, r=<optimized out>) at ../openssl-1.1.1d/crypto/x509/x_all.c:26
#7 0x0000ffffbf37f310 in internal_verify (ctx=0xaaaaaab50480) at ../openssl-1.1.1d/crypto/x509/x509_vfy.c:1750
#8 0x0000ffffbf380c44 in verify_chain (ctx=0xaaaaaab50480) at ../openssl-1.1.1d/crypto/x509/x509_vfy.c:232
#9 0x0000ffffbf381358 in X509_verify_cert (ctx=ctx@entry=0xaaaaaab50480) at ../openssl-1.1.1d/crypto/x509/x509_vfy.c:295
#10 0x0000ffffbf45516c in ssl_verify_cert_chain (s=s@entry=0xaaaaaab46150, sk=sk@entry=0xaaaaaab45a30) at ../openssl-1.1.1d/ssl/ssl_cert.c:427
#11 0x0000ffffbf475af4 in tls_process_server_certificate (s=0xaaaaaab46150, pkt=0xfffffffff258) at ../openssl-1.1.1d/ssl/statem/statem_clnt.c:1897
#12 0x0000ffffbf478014 in ossl_statem_client_process_message (s=0xaaaaaab46150, pkt=<optimized out>) at ../openssl-1.1.1d/ssl/statem/statem_clnt.c:1039
#13 0x0000ffffbf471f1c in read_state_machine (s=0xaaaaaab46150) at ../openssl-1.1.1d/ssl/statem/statem.c:636
#14 state_machine (s=0xaaaaaab46150, server=65535) at ../openssl-1.1.1d/ssl/statem/statem.c:434
#15 0x0000ffffbf45e5fc in SSL_do_handshake (s=0xaaaaaab46150) at ../openssl-1.1.1d/ssl/ssl_lib.c:3607
#16 0x0000ffffbf6af38c in ossl_connect_step2 () from target:/usr/lib64/libcurl.so.4
#17 0x0000ffffbf6b38e0 in ossl_connect_common () from target:/usr/lib64/libcurl.so.4
#18 0x0000ffffbf6b4a10 in Curl_ssl_connect_nonblocking () from target:/usr/lib64/libcurl.so.4
#19 0x0000ffffbf677bdc in https_connecting () from target:/usr/lib64/libcurl.so.4
#20 0x0000ffffbf693b38 in multi_runsingle () from target:/usr/lib64/libcurl.so.4
#21 0x0000ffffbf694aec in curl_multi_perform () from target:/usr/lib64/libcurl.so.4
#22 0x0000ffffbf68cf54 in curl_easy_perform () from target:/usr/lib64/libcurl.so.4
```

**Comment by mdurnev** [comment]
Could anyone look into this issue please?

**Comment by mdurnev** [comment]
Another solution is:
```
diff -Naurp a/src/p11_pkey.c b/src/p11_pkey.c
--- a/src/p11_pkey.c	2020-02-19 16:17:49.000000000 +1000
+++ b/src/p11_pkey.c	2020-02-27 11:38:36.545722279 +1000
@@ -77,6 +77,13 @@ struct evp_pkey_method_st {
 } /* EVP_PKEY_METHOD */ ;
 #endif
 
+#define EVP_PKEY_FLAG_DYNAMIC   1
+
+struct evp_pkey_method_header_st {
+    int pkey_id;
+    int flags;
+};
+
 #if OPENSSL_VERSION_NUMBER < 0x10002000L || defined(LIBRESSL_VERSION_NUMBER)
 
 typedef struct {
@@ -508,6 +515,9 @@ static EVP_PKEY_METHOD *pkcs11_pkey_meth
 	new_meth = EVP_PKEY_meth_new(EVP_PKEY_RSA,
 		EVP_PKEY_FLAG_AUTOARGLEN);
 
+	/* do not allow OpenSSL to free this object */
+	((struct evp_pkey_method_header_st*)new_meth)->flags &= ~EVP_PKEY_FLAG_DYNAMIC;
+
 	EVP_PKEY_meth_copy(new_meth, orig_meth);
 
 	EVP_PKEY_meth_set_sign(new_meth,
@@ -648,6 +658,9 @@ static EVP_PKEY_METHOD *pkcs11_pkey_meth
 	new_meth = EVP_PKEY_meth_new(EVP_PKEY_EC,
 		EVP_PKEY_FLAG_AUTOARGLEN);
 
+	/* do not allow OpenSSL to free this object */
+	((struct evp_pkey_method_header_st*)new_meth)->flags &= ~EVP_PKEY_FLAG_DYNAMIC;
+
 	EVP_PKEY_meth_copy(new_meth, orig_meth);
 
 	EVP_PKEY_meth_set_sign(new_meth,
```

But it directly modifies some OpenSSL data structures.

**Comment by mtrojnar** [comment]
I'm confused with the proposed fix.  Wasn't it already merged in https://github.com/OpenSC/libp11/commit/e64496a198d4d2eb0310a22dc21be8b81367d319?

**Comment by mdurnev** [comment]
@mtrojnar, thank you for your response. I worked with release 4.10, not with the master branch. So I did not notice your patch. Thanks for pointing that out. In fact, I re-created the same patch and tested it with some logging. It allocates a bit more memory than we actually need. It depends on how many times PKCS11_pkey_meths is called. That's why I proposed another patch. Instead of re-allocating memory on each call it protects the pointers from being freed (see my second patch in my previous comment). I am not an expert in OpenSSL, so I'd like to know your opinion on that. Which fix is better? Looking forward to receiving your feedback again.
